### PR TITLE
toc: modify disk format to include names for backwards/forwards compatibility

### PR DESCRIPTION
**Pull request by rmmh**
Including section names in the table of contents permits simpler
forwards and backwards compatibility. Instead of having to bump
the entire IndexFormatVersion when a new section is added, there is
simply a new section present in the table of contents.

Older versions can read TOCs with unknown sections and skip over them
with a warning. This is useful to permit downgrades without always
requiring a reindex.

Newer versions can read TOCs from older version with missing sections
and handle them gracefully, by checking for empty sections when loading
an index file and implementing whatever fallback code is necessary.

Section evolution is possible by having a new name for a tagged section,
adding the old section to the CompatibilityList, and writing the
conversion code when loading the file, or modifying the users of the
section to use whichever one is loaded.

Change-Id: I9aa05f29eb9d64fd0fff218f008d2031f1a15c8c

### CODE REVIEW

**Comment by rmmh** [comment]
This might break reading currently merged shards, but should make it simpler to handle the shard merging read/write code.

File: write.go [hunk 1]
```diff
@@ -25,10 +25,23 @@ import (
...
-func (w *writer) writeTOC(secs []section) {
-	w.U32(uint32(len(secs)))
+func (w *writer) writeTOC(toc *indexTOC) {
+	// Tagged sections are indicated with a 0 section count.
+	// Tagged sections allow easier forwards and backwards
+	// compatibility when evolving zoekt index files with new
+	// sections.
+	//
+	// A tagged section is:
+	// Varint TagLen, Tag String, Varint SecType, Section
```
Comment (stefanhengl) [inline write.go]:
```suggestion
	// Varint TagLen, Tag String, Varint Kind, Section
```

File: write.go [hunk 2]
```diff
@@ -25,10 +25,23 @@ import (
...
+	//
+	// Section type is indicated because simpleSections and
```
Comment (stefanhengl) [inline write.go]:
```suggestion
	// Section kind is indicated because simpleSections and
```

File: toc.go [hunk 3]
```diff
@@ -40,11 +40,31 @@ const IndexFormatVersion = 16
...
+// longer valid.
+// In this way, compatibility with arbitrary version offsets can be indicated.
+
+// MinReaderVersion constrains forwards compatibility by emitting files
+// that won't load in zoekt with a FeatureVersion below it.
+const MinReaderVersion = 10
+
+// ReaderMinFeatureVersion constrains backwards compatibility by refusing to
+// load a file with a FeatureVersion below it.
+const ReaderMinFeatureVersion = 8
```
Comment (stefanhengl) [inline toc.go]: nit: It was dificult for me to remember the difference between MinReaderVersion and ReaderMinFeatureVersion. I had to go back to this text several times.

What about renaming to something like this

FeatureVersion -> ReaderVersion
MinReaderVersion -> WriteMinReaderVersion
ReaderMinFeatureVersion -> ReadMinReaderVersion

**Review by stefanhengl — APPROVED** [review]
This looks great! I really like the new index format and the way we specify a bracket of supported feature versions.

For which cases do we still need FormatVersion in the future? It seems like we can always go the way of adding a new section and then deprecating the old one later, just like you describe in the docstrings.

File: write.go (hunk 1 above)
Comment (rmmh) [reply write.go]: Done.

File: write.go (hunk 2 above)
Comment (rmmh) [reply write.go]: Done.

File: toc.go (hunk 3 above)
Comment (rmmh) [reply toc.go]: I did the second two but not the first-- FeatureVersion is part of the metadata json and I think "FeatureVersion" is more descriptive than "ReaderVersion".

**Comment by rmmh** [comment]
I'm not sure what would necessitate a FormatVersion change in the future.

If we make enough section changes without a FormatVersion evolution, we might drop it altogether later.

**Pull request merged**